File: C++/4-Stack/Problems/1-ValidParentheses/ValidParenthesesPro.js
/**
 * 1. Stack Approach (Most Common)
 *
 * * Concept: Uses a stack to keep track of opening brackets.  When a closing bracket is encountered,
 * it checks if it matches the top of the stack.
 * * Real-world: Useful in compilers to validate code syntax, in XML/JSON parsers to check tag matching,
 * and in text editors for bracket matching.
 */
function isValidStack(s) {
    const stack = [];
    const map = {
        ')': '(',
        '}': '{',
        ']': '['
    };

    for (let i = 0; i < s.length; i++) {
        const char = s[i];

        if (map[char]) { // If it's a closing bracket
            const topElement = stack.pop() || '#'; // Use '#' as a dummy if stack is empty
            if (topElement !== map[char]) {
                return false;
            }
        } else { // If it's an opening bracket
            stack.push(char);
        }
    }

    return stack.length === 0;
}

// 2. Stack Approach with Early Exit
// Concept: Similar to the first approach, but adds an optimization to exit early if a mismatch is found.
function isValidStackEarlyExit(s) {
  const stack = [];
  const map = {
    ')': '(',
    '}': '{',
    ']': '['
  };

  for (let i = 0; i < s.length; i++) {
    const char = s[i];

    if (map[char]) {
      const topElement = stack.pop();
      if (!topElement || topElement !== map[char]) { // Check for empty stack *and* mismatch
        return false;
      }
    } else {
      stack.push(char);
    }
  }
  return stack.length === 0;
}

/**
 * 3. Recursive Approach (Less Common, but good for understanding recursion)
 *
 * * Concept: Recursively checks if the first bracket and its matching closing bracket
 * are valid, then recursively checks the rest of the string.
 * * Real-world:  While less common for this specific problem, recursion is fundamental
 * in many algorithms like tree traversals, graph searches, and parsing complex structures.  This
 * demonstrates a basic recursive pattern.
 */
function isValidRecursive(s) {
    if (s.length === 0) return true;

    const matching = { '(': ')', '{': '}', '[': ']' };
    for (let i = 0; i < s.length; i++) {
        if (matching[s[i]]) {
            let count = 1;
            for (let j = i + 1; j < s.length; j++) {
                if (s[j] === matching[s[i]]) {
                    count--;
                    if (count === 0) {
                        const restOfStringValid = isValidRecursive(s.substring(i + 1, j)) && isValidRecursive(s.substring(j+1));
                        return restOfStringValid;
                    }
                } else if (s[j] === s[i]) {
                    count++;
                }
            }
            return false; //closing bracket not found
        }
    }
    return false;
}

/**
 * 4. Replace and Check (Efficient for certain patterns)
 *
 * * Concept: Repeatedly removes all matching pairs of parentheses from the string.
 * If the string is empty at the end, it's valid.
 * * Real-world: Useful in some string processing scenarios where you need to simplify
 * or normalize strings by removing matching patterns.  Can be adapted for tasks
 * like removing redundant delimiters or simplifying code.
 */
function isValidReplace(s) {
    let originalLength;
    do {
        originalLength = s.length;
        s = s.replace(/\(\)|{}|\[\]/g, '');
    } while (s.length !== originalLength);
    return s.length === 0;
}

/**
 * 5. Using a Map and Single Loop (Optimized)
 *
 * Concept: Uses a map to store the opening-closing pairs and efficiently checks the string in a single loop.
 * Real-world: This approach is very efficient and can be used in performance-critical applications
 * such as validating large code files or processing a high volume of data with bracketed structures.
 */
function isValidMap(s) {
    if (s.length % 2 !== 0) return false; // Odd length strings cannot be valid.

    const map = new Map([
        [')', '('],
        ['}', '{'],
        [']', '[']
    ]);
    const stack = [];

    for (let i = 0; i < s.length; i++) {
        const char = s[i];
        if (map.has(char)) { // Closing bracket
            const top = stack.pop();
            if (top !== map.get(char)) {
                return false;
            }
        } else { // Opening bracket
            stack.push(char);
        }
    }
    return stack.length === 0;
}

// Test Cases
const testCases = [
    "()",
    "()[]{}",
    "(]",
    "([)]",
    "{[]}",
    "((()))",
    "[{()}]",
    "}",
    "(",
    "}{",
    "((",
    "))",
    "())(",
    "((())",
    "((",
    "))"
];

// Run tests and log results
console.log("--- Stack Approach ---");
testCases.forEach(testCase => {
    console.log(`Input: ${testCase}, Output: ${isValidStack(testCase)}`);
});

console.log("\n--- Stack Approach with Early Exit ---");
testCases.forEach(testCase => {
    console.log(`Input: ${testCase}, Output: ${isValidStackEarlyExit(testCase)}`);
});

console.log("\n--- Recursive Approach ---");
testCases.forEach(testCase => {
    console.log(`Input: ${testCase}, Output: ${isValidRecursive(testCase)}`);
});

console.log("\n--- Replace and Check Approach ---");
testCases.forEach(testCase => {
    console.log(`Input: ${testCase}, Output: ${isValidReplace(testCase)}`);
});

console.log("\n--- Map and Single Loop Approach ---");
testCases.forEach(testCase => {
    console.log(`Input: ${testCase}, Output: ${isValidMap(testCase)}`);
});

File: C++/4-Stack/Problems/1-ValidParentheses/ValidParenthesesPro.cpp
#include <iostream>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>
#include <algorithm>
#include <stdexcept> // For exception handling

using namespace std;

// Approach 1: Classic Stack Solution
//   - Uses a stack to track opening brackets and checks for proper closing.
//   - Suitable for basic validation in any application that processes bracketed input.
bool isValid_Stack(const string& s) {
    if (s.empty()) return true; // Empty string is considered valid
    if (s.size() % 2 != 0) return false; // Odd length strings cannot be valid

    stack<char> bracketStack;
    for (char c : s) {
        switch (c) {
            case '(': bracketStack.push(c); break;
            case '{': bracketStack.push(c); break;
            case '[': bracketStack.push(c); break;
            case ')':
                if (bracketStack.empty() || bracketStack.top() != '(') return false;
                bracketStack.pop();
                break;
            case '}':
                if (bracketStack.empty() || bracketStack.top() != '{') return false;
                bracketStack.pop();
                break;
            case ']':
                if (bracketStack.empty() || bracketStack.top() != '[') return false;
                bracketStack.pop();
                break;
            default:
                // Handle invalid characters.  In a real application, you might:
                // 1. Log an error.
                // 2. Throw an exception.
                // 3. Ignore the character.
                return false; // For this problem, we'll treat invalid chars as making the string invalid.
        }
    }
    return bracketStack.empty(); // The string is valid if the stack is empty at the end.
}

// Approach 2: Stack with Early Exit for Mismatched Counts
//   - Optimized version that checks counts of opening and closing brackets early on.
//   - Useful when dealing with potentially very long strings, providing a quick check before more intensive processing.
bool isValid_StackEarlyExit(const string& s) {
    if (s.empty()) return true;
    if (s.size() % 2 != 0) return false;

    int openParenCount = 0, openBraceCount = 0, openBracketCount = 0;
    for (char c : s) {
        switch (c) {
            case '(': openParenCount++; break;
            case '{': openBraceCount++; break;
            case '[': openBracketCount++; break;
            case ')': openParenCount--; break;
            case '}': openBraceCount--; break;
            case ']': openBracketCount--; break;
            default: return false;
        }
    }
    if (openParenCount != 0 || openBraceCount != 0 || openBracketCount != 0) return false;

    stack<char> bracketStack;
    for (char c : s) {
        switch (c) {
            case '(': bracketStack.push(c); break;
            case '{': bracketStack.push(c); break;
            case '[': bracketStack.push(c); break;
            case ')':
                if (bracketStack.empty() || bracketStack.top() != '(') return false;
                bracketStack.pop();
                break;
            case '}':
                if (bracketStack.empty() || bracketStack.top() != '{') return false;
                bracketStack.pop();
                break;
            case ']':
                if (bracketStack.empty() || bracketStack.top() != '[') return false;
                bracketStack.pop();
                break;
            default: return false;
        }
    }
    return bracketStack.empty();
}

// Approach 3: Using a Map for Matching
//   - Employs an unordered_map for efficient matching of closing to opening brackets.
//   - Improves readability and can be slightly faster for large inputs due to the hash table lookup.
bool isValid_Map(const string& s) {
    if (s.empty()) return true;
    if (s.size() % 2 != 0) return false;

    unordered_map<char, char> matchingPairs = {
        {')', '('},
        {'}', '{'},
        {']', '['}
    };
    stack<char> bracketStack;
    for (char c : s) {
        if (c == '(' || c == '{' || c == '[') {
            bracketStack.push(c);
        } else if (c == ')' || c == '}' || c == ']') {
            if (bracketStack.empty() || bracketStack.top() != matchingPairs[c]) {
                return false;
            }
            bracketStack.pop();
        } else {
            return false;
        }
    }
    return bracketStack.empty();
}

// Approach 4: Iterative Replacement (Not Recommended for Large Inputs)
//   - Repeatedly replaces matching pairs "()", "{}", and "[]" with empty strings.
//   - Simpler to understand but can be inefficient for long or complex strings due to repeated string operations.
//   - Best suited for very short strings or as a demonstration of a different approach.  Avoid in performance-critical code.
bool isValid_Replace(string s) {
    if (s.empty()) return true;
    if (s.size() % 2 != 0) return false;

    while (s.find("()") != string::npos || s.find("{}") != string::npos || s.find("[]") != string::npos) {
        size_t pos;
        pos = s.find("()");
        if (pos != string::npos) s.replace(pos, 2, "");
        pos = s.find("{}");
        if (pos != string::npos) s.replace(pos, 2, "");
        pos = s.find("[]");
        if (pos != string::npos) s.replace(pos, 2, "");
    }
    return s.empty();
}

// Approach 5: Recursive Solution (For Demonstration - Can Cause Stack Overflow)
//   - Recursively checks for and removes matching pairs.
//   - Elegant but can be very inefficient and lead to stack overflow for deeply nested brackets.
//   - Primarily for educational purposes, demonstrating recursion.  Avoid in production code for potentially large inputs.
bool isValid_Recursive(const string& s) {
    if (s.empty()) return true;
    if (s.size() % 2 != 0) return false;

    size_t pos;
    string modifiedS = s; // Create a copy to modify

    pos = modifiedS.find("()");
    if (pos != string::npos) {
        modifiedS.replace(pos, 2, "");
        return isValid_Recursive(modifiedS);
    }
    pos = modifiedS.find("{}");
    if (pos != string::npos) {
        modifiedS.replace(pos, 2, "");
        return isValid_Recursive(modifiedS);
    }
    pos = modifiedS.find("[]");
    if (pos != string::npos) {
        modifiedS.replace(pos, 2, "");
        return isValid_Recursive(modifiedS);
    }
    return modifiedS.empty();
}

int main() {
    // Test cases to demonstrate the different approaches
    vector<string> testCases = {
        "()",
        "()[]{}",
        "(]",
        "([)]",
        "{[]}",
        "",
        "((()))",
        "[{()}]",
        "((",
        "))",
        "}{",
        "[(])",
        "()[]{}(){}",
        "((()))[{}]",
        "{{{{}}}}"
    };

    cout << "Valid Parentheses - Test Cases and Results:\n";
    cout << "-------------------------------------------\n";

    for (const string& testCase : testCases) {
        cout << "Test Case: \"" << testCase << "\"\n";
        cout << "--------------------\n";
        try {
            cout << "Stack Approach:         " << (isValid_Stack(testCase) ? "Valid" : "Invalid") << "\n";
            cout << "Stack Early Exit:       " << (isValid_StackEarlyExit(testCase) ? "Valid" : "Invalid") << "\n";
            cout << "Map Approach:           " << (isValid_Map(testCase) ? "Valid" : "Invalid") << "\n";
            cout << "Replace Approach:       " << (isValid_Replace(testCase) ? "Valid" : "Invalid") << "\n";
            cout << "Recursive Approach:     " << (isValid_Recursive(testCase) ? "Valid" : "Invalid") << "\n";
        } catch (const std::exception& e) {
            cerr << "An exception occurred: " << e.what() << "\n";
        }
        cout << "\n";
    }

    // Example of using the isValid function in a real-world scenario (e.g., config file validation)
    cout << "\nReal-world Example: Configuration File Validation\n";
    cout << "--------------------------------------------------\n";
    string configFile = "[{ \"server\": \"localhost\", \"port\": 8080, \"connections\": (100) }]";
    if (isValid_Stack(configFile)) { // Use the Stack approach, as it's generally the most robust
        cout << "Configuration file is valid.\n";
        //  Parse the configuration file here (e.g., using a JSON parser).  This is where you'd
        //  integrate this validation into your application's startup or configuration loading.
    } else {
        cout << "Configuration file is invalid: " << configFile << "\n";
        //  Handle the error:  log it, display an error message to the user, or exit the application.
    }

    cout << "\nReal-world Example: Code Editor/IDE Bracket Matching\n";
    cout << "--------------------------------------------------\n";
    string codeSnippet = "function example(arg1: string, arg2: number[]): void { console.log(arg1, arg2); }";
    if(isValid_Stack(codeSnippet)) {
        cout << "Code snippet has balanced brackets.\n";
    }
    else {
        cout << "Code snippet has unbalanced brackets.\n";
    }

    return 0;
}
